File: CS12B/lab5/Dictionary.c
//------------------------------------------------------------------------------
// Sudeep Baniya
// CMPS 12M
// sucbaniy
// 5-14-2016
// The programs implements a Dictionary ADT based on the linked list data strucure// Dictionary.c
//-------------------------------------------------------------------------------

#include<stdio.h>

#include<stdlib.h>

#include<string.h>

#include<assert.h>

#include"Dictionary.h"


// private types --------------------------------------------------------------

// NodeObj

typedef struct NodeObj{

   char* value;

   char* key;

   struct NodeObj* next;

} NodeObj;

// Node
typedef NodeObj* Node;

// newNode()
// constructor of the Node type
Node newNode(char* k, char* v) {

   Node N = malloc(sizeof(NodeObj));

   assert(N!=NULL);

   N->value = v;

   N->key = k;

   N->next = NULL;

   return(N);

}

// freeNode()
// destructor for the Node type
void freeNode(Node* pN){

   if( pN!=NULL && *pN!=NULL ){

      free(*pN);

      *pN = NULL;

   }
}

// DictionaryObj
typedef struct DictionaryObj{

   Node first;

   int numOfItems;

} DictionaryObj;

//helper fuctions for delete()
void deleteEverything(Node N){

  if (N!=NULL){

   deleteEverything(N->next);

    freeNode(&N);
  }
}

//findKey()
//this returns reference to the NodeObject with its argument key or else return null if no NodeObject exits
Node findKey(Node A, char* targetKey){
 
 while(A != NULL){
 
    if(strcmp(A->key, targetKey)==0){
 
      break;
    }
 
    A = A->next;

    }

    return A;
}

//helper functions for printDictionary()
void stringout(FILE* out, Node A){

  if(A!=NULL){

    stringout(out, A->next);

    fprintf(out, "%s %s\n", A->key, A->value);
  }
}


// public functions -----------------------------------------------------------

// newDictionary()
// constructor for the Dictionary type
Dictionary newDictionary(void){

   Dictionary D = malloc(sizeof(DictionaryObj));

   assert(D!=NULL);

   D->first = NULL;

   D->numOfItems = 0;

   return D;
}

// Dictionary()
// destructor for the Dictionary type
void freeDictionary(Dictionary* pD){

   if( pD!=NULL && *pD!=NULL ){

      if( !isEmpty(*pD) ) makeEmpty(*pD);

      free(*pD);

      *pD = NULL;

   }
}

// isEmpty()
// returns 1 (true) if S is empty, 0 (false) otherwise
// pre: none
int isEmpty(Dictionary D){

   if( D==NULL ){

      fprintf(stderr, 

              "Dictionary Error: calling isEmpty() on NULL reference\n");

      exit(EXIT_FAILURE);

   }

   return(D->numOfItems==0);

}

//size()
// returns the number of (key, value) pairs in D
// pre none
int size(Dictionary D){

  return D->numOfItems;
}

// lookup()
// returns the value v such that (k,v) is in D, or returns NULL if no such value v exists
// pre: none
char* lookup(Dictionary D, char* k){

  Node N = findKey(D->first, k);

  if(N==NULL){

    return NULL;

  }else{
   
    return N->value;
  }
}

// insert()
// inserts new (key, value) pair into D
// pre: lookup(D, k)==NULL
void insert(Dictionary D, char* k, char* v){
  
  Node N;  
  
  if(findKey(D->first, k)!=NULL){
  
    fprintf(stderr, 
      "Dictionary Error: calling insert() when key exists\n");
  
    exit(EXIT_FAILURE);
  
  }
  
  N = newNode(k, v);
  
  N->next = D->first;
  
  D->first = N;
  
  N = NULL;
  
  D->numOfItems++;
}

// delete()
// deletes pair with the key k
// pre: lookup(D, k)!=NULL
void delete(Dictionary D, char* k){

  Node N;

  if(findKey(D->first, k)==NULL){

    fprintf(stderr, 
      "Dictionary Error: calling delete() when nothing exits\n");

    exit(EXIT_FAILURE);

  }

  if(findKey(D->first, k)==D->first){

    N = D->first;

    D->first = D->first->next;

    N->next = NULL;

  }else{

    N = findKey(D->first, k);

    Node before = D->first;

    Node temporary = D->first->next;

    while(temporary !=N){

      temporary = temporary->next;

      before = before->next;
    }
    
    before->next = N->next;
    
    N->next = NULL;
  }
  
  D->numOfItems--;
  
  freeNode(&N);

}

// makeEmpty()
// resets D to the empty state
// pre: !isEmpty(D)
void makeEmpty(Dictionary D){
  
   if( D==NULL ){
  
      fprintf(stderr, 
  
              "Dictionary Error: calling makeEmpty() on NULL Dictionary reference\n");
  
      exit(EXIT_FAILURE);
  
   }
  
   if( D->numOfItems==0 ){
  
      fprintf(stderr, 
              "Dictionary Error: calling makeEmpty() on empty Dictionary\n");
  
      exit(EXIT_FAILURE);
   }
  
   deleteEverything(D->first);
  
   D->first = NULL;
  
   D->numOfItems = 0;
}

//printDictionary()
//pre: none
//prints a text representation of D to the file pointed to by out
void printDictionary(FILE* out, Dictionary D){
  
   if( D==NULL ){
  
      fprintf(stderr, 
              "Dictionary Error: calling printDictionary() on where the reference is NULL\n");
  
      exit(EXIT_FAILURE);
  
   }
  
   stringout(out, D->first); 
}

File: CS12B/pa3/Dictionary.java
//------------------------------------------------------------------------------
// Sudeep Baniya
// CMPS 12B
// sucbaniy
// 4-28-2016
// The programs implements a Dictionary ADT based on the linked list data strucure.
// Dictionary.java
//------------------------------------------------------------------------------
public class Dictionary implements DictionaryInterface {

  // private inner Node class
  private class Node {
    Node next;
    String key;
    String value;

    // node class constructor
    Node(String key, String value) {
      next = null;
      this.key = key;
      this.value = value;
    }
  }

  // Fields for the Dictionary class
  private Node first;      // refers to first Node in Dictionary
  private int numItems;   // the number of items in this Dictionary

  // Dictonary()
  // constructor for the Dictionary class
  public Dictionary() {
    first = null;
    numItems = 0;
  }

  // helper functions -------------------------------------------------

   // findKey()
   // returns a reference to the Node at containing its argument key or else return null
   private Node findKey(String key) {
     Node N = first;
     while (N != null){
       if(N.key.equals(key)){
         break;
       }else{ 
          N = N.next;
       }
   }
    return N;
}
  
  // below contains isEmpty(),size(),lookup(),insert(),delete(),makeEmpty(),toString()

   // isEmpty()
   // pre: none
   // returns true if dictionary contains no pairs and returns falls other wise
   public boolean isEmpty() {
     return (numItems==0);
   }

   // size()
   // pre: none
   // returns the number of entries in the dictionary
   public int size() {
     return numItems;
   }

   // lookup()
   // pre: none
   // if contains a pair who key machtes the argument key, it returns the field else returns null
   public String lookup(String key) {
     Node N = first;
     while (N != null){
       if (N.key.equals(key)){
         return N.value;
       }
       N = N.next;
     }
     return null;
   }

   // insert()
   // inserts a new (key,value) if it is still not int the list
   // pre: lookup(key)==null
   public void insert(String key, String value) 
   throws DuplicateKeyException {
     if (lookup(key) != null){
       throw new DuplicateKeyException("cannot insert duplicate keys");
     }else{
      if( first == null){
        Node N = new Node(key,value);
        first = N;
        numItems++;
      }else{
        Node N = first;
        while( N != null){
        if(N.next == null){
        break;
        }
        N = N.next;
        }
        N.next = new Node(key,value);
        numItems++;
      }
    }
  }

  // delete()
  // if the pair matches the key then it is deleted from the dictionary
  // pre: lookup(key)!=null
  public void delete(String key) throws KeyNotFoundException {
    
    if( lookup(key) == null){
      throw new KeyNotFoundException("cannot delete non-existent key");
    } else {
      if(findKey(key) == first){
      Node N = first;
      first = first.next;
      N.next = null;
    }else{
      Node N = findKey(key);
      Node back = first;
      Node temporary = first.next;
      while (temporary != N){
        temporary = temporary.next;
        back = back.next;
      }
      back.next = N.next;
      N.next = null;
    }
    numItems--;
  }
}

  // makeEmpty()
  // pre: none
  // return the dictionary to have nothing
  public void makeEmpty(){
    first = null;
    numItems = 0;
  }

  // toString()
  // returns a String representation of this Dictionary
  // overrides Object's toString() method
  // pre: none
  public String toString() {
    StringBuffer sb = new StringBuffer();
    Node N = first;
    for(N = first; N != null; N = N.next){ 
    sb.append(N.key).append(" ").append(N.value).append("\n");
  }
  return new String(sb);
 }
}
